File: pegasusX/packages/mobile-ios-barcode/EANBarcode.swift
//
//  EANBarcode.swift
//  Shared EAN/GTIN normalization for catalog and return-gate flows.
//

import Foundation

enum EANBarcode {
    /// Strips non-digits and validates GTIN length + checksum (EAN-8/12/13/14).
    static func normalize(_ raw: String) -> String? {
        let digits = raw.filter(\.isNumber)
        guard !digits.isEmpty else { return nil }
        guard [8, 12, 13, 14].contains(digits.count) else { return nil }
        guard validGtinChecksum(digits) else { return nil }
        return digits
    }

    private static func validGtinChecksum(_ code: String) -> Bool {
        let chars = Array(code)
        guard chars.count >= 8 else { return false }
        var sum = 0
        let n = chars.count
        for i in 0..<(n - 1) {
            guard let d = chars[i].wholeNumberValue else { return false }
            let posFromRight = n - 1 - i
            sum += (posFromRight % 2 == 1) ? d * 3 : d
        }
        guard let check = chars[n - 1].wholeNumberValue else { return false }
        return (10 - (sum % 10)) % 10 == check
    }
}

File: pegasusX/packages/mobile-android-barcode-scanner/src/main/java/com/pegasus/barcode/EanBarcode.kt
package com.pegasus.barcode

/** Normalizes retail EAN/GTIN input (digits only, valid length + checksum). */
object EanBarcode {
    fun normalize(raw: String): String? {
        val digits = raw.filter { it.isDigit() }
        if (digits.isEmpty()) return null
        if (digits.length !in setOf(8, 12, 13, 14)) return null
        if (!validGtinChecksum(digits)) return null
        return digits
    }

    private fun validGtinChecksum(code: String): Boolean {
        val n = code.length
        if (n < 8) return false
        var sum = 0
        for (i in 0 until n - 1) {
            val d = code[i].digitToIntOrNull() ?: return false
            val posFromRight = n - 1 - i
            sum += if (posFromRight % 2 == 1) d * 3 else d
        }
        val check = code.last().digitToIntOrNull() ?: return false
        return (10 - (sum % 10)) % 10 == check
    }
}
